File: languages/haskell/01-basics-types-functions/src/control_flow.hs
-- Exercise 2: Control Flow — Pattern Matching and Guards
-- Compile: ghc control_flow.hs -o control_flow && ./control_flow

-- Pattern matching on function arguments
grade :: Int -> String
grade score
    | score >= 90 = "A"
    | score >= 80 = "B"
    | score >= 70 = "C"
    | otherwise   = "F"

-- Pattern matching on values
dayName :: Int -> String
dayName 1 = "Monday"
dayName 2 = "Tuesday"
dayName 3 = "Wednesday"
dayName 4 = "Thursday"
dayName 5 = "Friday"
dayName 6 = "Saturday"
dayName 7 = "Sunday"
dayName _ = "Unknown"

-- Recursive countdown with pattern matching
countDown :: Int -> IO ()
countDown 0 = putStrLn "Go!"
countDown n = do
    putStr (show n ++ "... ")
    countDown (n - 1)

main :: IO ()
main = do
    let score = 87
    putStrLn $ "Score: " ++ show score ++ " -> Grade: " ++ grade score

    let day = 3
    putStrLn $ "Day " ++ show day ++ " is " ++ dayName day

    putStr "Counting 1 to 5: "
    sequence_ [putStr (show i ++ " ") | i <- [1..5]]
    putStrLn ""

    putStrLn "Countdown: "
    countDown 5

File: languages/swift/01-basics-variables-types-control-flow/Sources/Basics/control_flow.swift
// Exercise 2: Control Flow — if, switch, for-in, while
import Foundation

func demonstrateControlFlow() {
    let score = 87

    // if-else
    let grade: String
    if score >= 90 {
        grade = "A"
    } else if score >= 80 {
        grade = "B"
    } else {
        grade = "C"
    }
    print("Score: \(score) → Grade: \(grade)")

    // switch — exhaustive
    let day = 3
    let dayName: String
    switch day {
    case 1:
        dayName = "Monday"
    case 2:
        dayName = "Tuesday"
    case 3:
        dayName = "Wednesday"
    case 4:
        dayName = "Thursday"
    case 5:
        dayName = "Friday"
    case 6, 7:
        dayName = "Weekend"
    default:
        dayName = "Unknown"
    }
    print("Day \(day) is \(dayName)")

    // for-in with ranges
    print("Counting 1 to 5: ", terminator: "")
    for i in 1...5 {
        print("\(i) ", terminator: "")
    }
    print()

    // while loop
    var countdown = 5
    while countdown > 0 {
        print("\(countdown)...", terminator: "")
        countdown -= 1
    }
    print("Go!")
}

demonstrateControlFlow()
